prql-compiler/src/semantic/transforms.rs

```diff
@@ -930,18 +932,34 @@ mod from_text {
             row.into_iter().map(|x| x.to_string()).collect()
         }
 
-        fn parse_row(row: csv::StringRecord) -> Vec<Literal> {
+        fn parse_row(row: csv::StringRecord) -> Result<Vec<Literal>> {
             row.into_iter()
-                .map(|x| Literal::String(x.to_string()))
-                .collect()
+                .map(|x| {
+                    // We use our own parser here, but possibly the code is no
+                    // simpler than if we wrote a mini-parser to understand
+                    // whether we need to parse a string or a number.
+                    Ok(crate::parser::parse(x)?
+                        .into_iter()
+                        .exactly_one()?
+                        .kind
+                        .as_main()
+                        .context("Expected main expr")?
+                        .as_ref()
+                        .kind
+                        .as_literal()
+                        .context("Expected literal expr")?
+                        .clone())
+                })
+                .try_collect()
         }
 
         Ok(RelationLiteral {
             columns: parse_header(rdr.headers()?),
             rows: rdr
                 .records()
                 .into_iter()
-                .map(|row_result| row_result.map(parse_row))
+                // .map(|row_result| row_result.map(|r| r.map(parse_row)))
+                .map(|row_result| row_result.map(parse_row)?)
```

Comment:
I'm probably wrong here but...could this become a `flat_map()`? 🤔

Reply:
Yes, I think that's right — feel very feel to have a go...

Reply:
I don't think so.

Type here would be `Iterator<Result<Row>>`. 

We want to return an error if any of the rows errors out. If I understand correctly, you'd use `.flat_map` to convert result into iter. That would convert `Result::Err` into empty iterator and just suppress the error.

Reply:
(good point, @aljazerzen is correct)